htmresearch/algorithms/column_pooler.py

```diff
@@ -296,34 +302,44 @@ def _chooseCells(self, candidates, n, numActiveSegmentsByCell):
     @return (list) Cells to activate
     """
 
-    orderedCandidates = sorted(candidates,
-                               key=numActiveSegmentsByCell.__getitem__,
-                               reverse=True)
-    activeSegmentCounts = sorted(set(numActiveSegmentsByCell[cell]
-                                     for cell in candidates),
-                                 reverse=True)
+    if n <= 0:
+      return numpy.empty(0)
+
+    # Keep a list of numpy arrays.
+    allChosenCells = []
+    numChosen = 0
 
-    chosenCells = []
-    i = 0
+    # Walk the active segment counts in decreasing order.
+    activeSegmentCounts = numpy.unique(
+      numActiveSegmentsByCell[candidates])[::-1]
 
     for activeSegmentCount in activeSegmentCounts:
-      if len(chosenCells) >= n or i >= len(orderedCandidates):
-        break
 
       if activeSegmentCount == 0:
-        chosenCells.extend(orderedCandidates[i:])
+        allChosenCells.append(candidates)
+        numChosen += candidates.size
         break
+      else:
+        # Activate all cells that are not inhibited by this segment count.
+        # A cell is inhibited if another cell has at least
+        # 'distalSegmentInhibitionFactor' times as many active segments.
+        boundary = float(activeSegmentCount) / self.distalSegmentInhibitionFactor
```

Comment:
Unlikely, but if a user specifies `distalSegmentInhibitionFactor=0` in constructor, is there any risk of a divide-by-zero here?  Would it be beneficial to provide some checks in the constructor?